src/history/hs_rec.c

```diff
@@ -511,8 +511,13 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
              * No need to continue if we found a first self contained value that is globally
              * visible.
              */
-            if (__wt_txn_upd_visible_all(session, upd) && WT_UPDATE_DATA_VALUE(upd))
+            if (__wt_txn_upd_visible_all(session, upd) && WT_UPDATE_DATA_VALUE(upd)) {
+                /* Free obsolete updates if exist. */
+                if (FLD_ISSET(S2C(session)->strategy_flags, WT_CONN_STRATEGY_OBSOLETE_CHECK) &&
+                  !F_ISSET(r, WT_REC_EVICT) && upd->next != NULL)
+                    __wt_free_update_next_list(session, upd);
```

Comment:
I did like the code that had a single way to identify obsolete updates - it's more subtle than ideal, and the free semantic requires care.

What was your motivation for going this other way? Is it to avoid re-traversing an update chain?

Reply:
Yes, it is to avoid re-traversing the update chain.